api/src/backend/tasks/tasks.py

```diff
@@ -152,3 +168,116 @@ def perform_scan_summary_task(tenant_id: str, scan_id: str):
+
+
+@shared_task(base=RLSTask, name="scan-output", queue="scans")
+@set_tenant(keep_tenant=True)
+def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
+    """
+    Process findings in batches and generate output files in multiple formats.
+
+    This function retrieves findings associated with a scan, processes them
+    in batches of 50, and writes each batch to the corresponding output files.
+    It reuses output writer instances across batches, updates them with each
+    batch of transformed findings, and uses a flag to indicate when the final
+    batch is being processed. Finally, the output files are compressed and
+    uploaded to S3.
+
+    Args:
+        tenant_id (str): The tenant identifier.
+        scan_id (str): The scan identifier.
+        provider_id (str): The provider_id id to be used in generating outputs.
+    """
+    # Generate and ensure the output directory exists
+    output_directory = _generate_output_directory(provider_id, tenant_id, scan_id)
+    os.makedirs("/".join(output_directory.split("/")[:-1]), exist_ok=True)
+
+    # Define auxiliary variables
+    output_writers = {}
+    scan_summary = FindingOutput._transform_findings_stats(
+        ScanSummary.objects.filter(scan_id=scan_id)
+    )
+
+    # Retrieve findings queryset
+    findings_qs = Finding.objects.filter(scan_id=scan_id).order_by("uid")
+
+    # Process findings in batches
+    for batch, is_last_batch in batched(findings_qs.iterator(), 50):
+        finding_outputs = [
+            FindingOutput.transform_api_finding(finding) for finding in batch
+        ]
+
+        # Generate output files
+        for mode, config in OUTPUT_FORMATS_MAPPING.items():
+            kwargs = dict(config.get("kwargs", {}))
+            if mode == "html":
+                kwargs["provider"] = initialize_prowler_provider(
+                    Provider.objects.get(id=provider_id)
+                )
+                kwargs["stats"] = scan_summary
+
+            writer_class = config["class"]
+            if writer_class in output_writers:
+                writer = output_writers[writer_class]
+                writer.transform(finding_outputs)
+                writer.close_file = is_last_batch
+            else:
+                writer = writer_class(
+                    findings=finding_outputs,
+                    file_path=output_directory,
+                    file_extension=config["suffix"],
+                )
+                writer.close_file = is_last_batch
+                output_writers[writer_class] = writer
+
+            # Write the current batch using the writer
+            writer.batch_write_data_to_file(**kwargs)
+            writer._data = []
+
+    # Compress output files
+    output_directory = _compress_output_files(output_directory)
+
+    # Save to configured storage
+    uploaded = _upload_to_s3(tenant_id, output_directory, scan_id)
+
+    if uploaded:
+        output_directory = uploaded
+        uploaded = True
```

Comment:
We need to add a cleanup if using S3 to remove files from `/tmp`

Reply:
Didn't we agree that the AWS machines are ephemeral so we don't mind deleting anything and locally it is inside the temporary files folder?

Reply:
Sure, but then I realised that the exports are going to run in a different queue, so it won't be ephemeral.